### ISSUE TITLE
Make the website accessible without javascript

### ISSUE THREAD

**Issue by purcell** [melpa/melpa#3728]
Right now the MELPA site is a javascript single page application, built from JSON data which is updated on the server side every few seconds, as the 3000+ packages are each built.

This allows the site to always reflect the server-side state accurately without requiring complex server-side machinery, but it means the site is entirely inaccessible to browsers without Javascript support. (It also makes it harder for us to obtain good website analytics data, but that's a lesser concern.)

We would like to investigate options for improving this situation. We would prefer to not require any long-running programs serving dynamically-generated pages to be running on our servers. Additionally, there should be a very minimal delay between packages being built and their status being reflected on the website. And any software we use should be as minimal as possible.

User "artlogic" suggested in [this Emacs Reddit thread](https://www.reddit.com/r/emacs/comments/4d0r7c/rms_there_is_a_big_ethical_problem_with_site/) that incremental rebuilds with a static site generator might be a viable option, particularly if we use Javascript enhancement for some of the more complex content.

Let's discuss options here and see if we can find a way ahead.

**Comment by artlogic** [comment]
Thanks for making this issue @purcell! Would you be willing to discuss a bit of how the current MELPA infrastructure is running? Are there any scripts that aren't in this repo? If I wanted to set it up myself, how would I do that? I have some ideas on how to move forward already, but it would be a lot easier to reason about MELPA if I understood what's happening server side.

**Comment by purcell** [comment]
Everything is in the repo, and it's easy to run locally, at least in its current form.

You can bootstrap a clone of the MELPA site locally by jumping into the `html` dir and running `make livejson`, then running `python -m SimpleHTTPServer` or `live-server` (from npm) in that dir.

On the server, the `recipes.json` file which lists all the recipes is assembled by code in `package-build.el`. And the large `archive.json` file is a JSON version of the standard `package.el`-format `archive-contents` file: it is also generated by `package-build.el`. The root `Makefile` triggers both these generation processes.

So you can essentially use `make` locally to duplicate what's happening on the server, with the only exception being the `download_counts.json` file, which is built from our web server access logs.

**Comment by artlogic** [comment]
Am I correct in thinking that the proper way forward would be to cause `make` (run against the root) to generate a static site in `html`?

**Comment by purcell** [comment]
Yeah, I think we'd need to make sure that we could render the entire site with a single `make` target, and also that it would be incrementally updated in response to running `make recipes/...`, for example.

I guess a naïve starting point would be to generate static html from the existing json.

**Comment by jwiegley** [comment]
Hi @purcell, I'm also interested in seeing a Javascript-free version, so that we can continue to reference MELPA from the documentation that ships with Emacs.

**Comment by artlogic** [comment]
@purcell @jwiegley I'm going to see about putting together a proof of concept in the next day or so using `erb` and `make` since they are already being used - unless someone else would rather work on it?

**Comment by dunn** [comment]
Re: license compatibility with Emacs, can the JS be explicitly licensed as GPL and marked in a LibreJS-compatible way?

**Comment by purcell** [comment]
> Re: license compatibility with Emacs, can the JS be explicitly licensed as GPL and marked in a LibreJS-compatible way?

@dunn Yes, I think that would be good. Any idea what the simplest way to accomplish that would be? And would switching our third-party scripts on cdnjs from minified to unminified code help?

**Comment by purcell** [comment]
> @purcell @jwiegley I'm going to see about putting together a proof of concept in the next day or so using erb and make since they are already being used - unless someone else would rather work on it?

That would be awesome, thanks.

**Comment by purcell** [comment]
> Hi @purcell, I'm also interested in seeing a Javascript-free version, so that we can continue to reference MELPA from the documentation that ships with Emacs.

By "Javascript-free", presumably you mean a version which mostly works without Javascript enabled, or do you mean that any Javascript served by the site would need to be deemed acceptable by LibreJS in any case?

**Comment by artlogic** [comment]
I'm thinking there are two issues at play here - one about reducing the dependence on Javascript and one about licensing.

Obviously, the site specific JS could be given a proper LibreJS tag: https://www.gnu.org/software/librejs/manual/html_node/Setting-Your-JavaScript-Free.html

I'm not sure if LibreJS is smart enough to notice that Mithril and moment.js are MIT Licensed (headers exist in their minified source code). Lodash is also MIT licensed, but refers to it's website instead of stating it outright. Cookies.js has no license header and appears to be public domain - see: https://github.com/ScottHamper/Cookies/blob/master/UNLICENSE (I realize both the "MIT License" and public domain are contentious topics).

I'm hoping to make most of these dependencies progressive enhancement, but it would still be nice to understand what LibreJS will do in the above cases. Does anyone know?

It might be a good idea to make a separate issue on JS licensing.

**Comment by purcell** [comment]
> It might be a good idea to make a separate issue on JS licensing.

I'll do that.

**Comment by jasonm23** [comment]
@artlogic you can assign tasks to me if you need any help (*instant_sunshine* on reddit.)

Also happy to look over any PRs if you want peer review.

**Comment by artlogic** [comment]
@purcell - one more question. Is it a safe assumption to make that the main site is simply running `make` from the project root in a loop? Is there a script that's driving that?

**Comment by purcell** [comment]
On the server it's run by [this runit script](https://github.com/melpa/melpa/blob/master/service/builder/run), which in turn calls [parallel_build_all](https://github.com/melpa/melpa/blob/master/scripts/parallel_build_all) -- note that this runs `make` with `-j4`, so we need to make sure each package build target (i.e. `recipes/*`, iirc) is safely parallelisable.

**Comment by milkypostman** [comment]
it's actually [syncer](https://github.com/melpa/melpa/blob/master/service/syncer/run) script which updates the website. every 5 minutes we run cleanup which does the following,

- look at all packages and remove any extra archives. i.e., if there is a new version of `evil` then delete the old version.
- update `archive-contents` to contain the newest versions of packages
- generate a new JSON file
- generate any new html

but these scripts are only running make in different ways.

as @purcell said, the `parallel_build_all` is running async to the updates and for each package it builds the archive and a metadata file which is eventually what gets put in `archive-contents` by the `syncer` script.

**Comment by artlogic** [comment]
@milkypostman Thanks. I had figured a fair amount of that out after @purcell's last comment, but it's great to have confirmation. Given that most of the data driving the website (archive.json and recipes.json) is updated by `syncer`, I'm going drive the proof of concept off that process and simply spit out a brand new site every 5 minutes. If that ends up being too slow, I'll look at doing incremental builds based on the individual recipes.

**Comment by milkypostman** [comment]
It may be better to have the recipe builder spit out the individual page
for each package. Then regen the main index page every 5 min.

**Comment by haight6716** [comment]
This seems more serious than you think - it looks like the reason that loading from melpa within emacs is entirely broken.  You need a non-js version of the site for melpa itself!  How is it supposed to get the package index if it's hidden in a json file that gets loaded by ajax?

**Comment by milkypostman** [comment]
Julian I don't understand what you're saying. Could you please try
explaining again?

**Comment by haight6716** [comment]
Sorry, I meant to refer to issue #3914 

In emacs, I'm getting this trying to install fiplr:
Contacting host: stable.melpa.org:443
package-handle-response: Error during download request: Not Found

**Comment by purcell** [comment]
@haight6716 That's entirely unrelated to this issue, sorry. Based on that error message you have now posted, your package list was out-of-date, and so your Emacs was trying to download an obsolete and no-longer-hosted package: use `package-refresh-contents` and then try again.

**Comment by haight6716** [comment]
@purcell  Thanks for the help and ( @milkypostman ) sorry for the false alarm!

**Comment by hpgisler** [comment]
Sorry for necrobumping...
Has there been any action on this recently?

(just being curious, as I've stumbled over this while EWW browsing melpa - which of course didn't work and which astonished me... ;-) )

**Comment by alanpearce** [comment]
So I'm working on this at the moment, there's a prototype up at https://melpa.alanpearce.eu which is working from [alanpearce/melpa#erb](https://github.com/alanpearce/melpa/tree/erb). 
I haven't **yet** implemented _needed by_ or some minor things like number beautification.
There's no search any more, so the index just displays the first ten packages. It would probably make more sense to have the ten latest updates, like the RSS feed does, or the top ten by downloads.

Right now it's using `erb` for templating, whilst `miller` does some pre-processing of the json files to split up archive.json into one file per package. I'm thinking that it would probably make more sense to get this to happen inside package-build, but I haven't looked into it just yet.

The templating is kind of slow though, even with split archive files. It takes around 5 minutes on a small Hetzner CPX11 VM with `make -j4`. I'm about to look into replacing the `erb` command line with [jeremyevans/erubi](https://github.com/jeremyevans/erubi), which is what Rails uses, apparently. I'm also thinking of experimenting with Caddy server's built-in templates.

[Hugo's data templates feature](https://gohugo.io/templates/data-templates/) is not suitable at the moment as it [doesn't support generating individual pages from a data source](https://github.com/gohugoio/hugo/issues/5074)

I would posit that we will need some javascript to redirect links from `/#/${PACKAGE}` to `/packages/${PACKAGE}.html`, otherwise people following old links will just end up on the homepage and with no search, that seems extra frustrating.

Curious to hear what you all think

**Comment by alanpearce** [comment]
Oh, and it's not at all integrated with the docker setup as I couldn't get that working. Some of the docker tags don't even exist any more, so they can't be pulled. Then even after fixing that as minimally as possible, I couldn't get things working well on macOS, so that's why my prototype is using a nix flake (of course I'll take this out when it's ready).

**Comment by hpgisler** [comment]
Just my insignificant small feedback on https://melpa.alanpearce.eu/ :+1: 
Exactly what I wish : simple, clean, clear and usable in eww.

**Comment by alanpearce** [comment]
I realised that I perhaps wasn't clear what feedback I was looking for in my last messages. 

1. Which approach would be preferred generally, or are there perhaps questions about them? nginx -> caddy + templating vs erb/erubi (despite slowness)
2. Should I implement search with javascript, or otherwise perhaps `all-packages.html`
3. Should I add `/#/foo` -> `/package/foo` redirects (only doable in JS)

Tagging @milkypostman since you mentioned having to set up a new server: probably you should have the biggest say on the first point.

**Comment by milkypostman** [comment]
@alanpearce I saw some CLs but I'm a bit overloaded right now with personal stuff. I'll keep this in mind and if you don't hear from me in a month can you ping me again?

**Comment by alanpearce** [comment]
No problem. Personal stuff always comes first!
